src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelThread.cs

```diff
@@ -262,11 +267,22 @@ private void ThreadStart(object parameter)
 
         private void OnPost()
         {
-            DoPostWork();
-            DoPostCloseHandle();
+            var loopsRemaining = _maxLoops;
+            bool wasWork;
+            do
+            {
+                // Queues swapped
+                wasWork = DoPostWork();
+                wasWork = DoPostCloseHandle() || wasWork;
```

Comment:
I started with that; then changed to this, didn't add reason :(

I think it was to make sure a minimum of a double loop had always been done and both queues processed in case a task had snuck in; to get the tasks into the Libuv queue as early as possible for lower latency. 

![libuv-loop](http://aoa.blob.core.windows.net/aspnet/loop_iteration_small.png)

But if it was `DoPostWork` that added an item to the queue then `wasWork` should return `true` so that would be ok as it would do another loop. If `wasWork` returned false, then `DoPostWork` and `DoPostCloseHandle` would both have returned pretty fast so probably wouldn't be anything extra there?

My worry was the libuv loop sleeping (15ms?) because it had nothing to do; however an item sitting in the thread queue ready to go; but missed due to `uv_async_send` [coalescing](http://docs.libuv.org/en/latest/async.html#c.uv_async_send) and blocked on the add behind the lock; but I was probably being over cautious?